### ISSUE TITLE
unable to get correct nyc/istanbul coverage report for typescript

### ISSUE THREAD

**Issue by ciekawy** [istanbuljs/nyc#618]
Please use the template provided below, when reporting bugs:
-----

## Expected Behavior
get correct coverage report for typescript project

## Observed Behavior
coverage is not correct for typescript files. i.e. html coverage map is miss-aligned with the source also coverage numbers seem to not be correct (and sometimes non deterministic

I've described my issue in https://stackoverflow.com/questions/44701991/getting-nyc-istanbul-coverage-report-to-work-with-typescript

copy-pasting  the info here:

`nyc` relevant config in `package.json`
```json
"scripts": {
    "test:coverage": "nyc npm run test:unit",
    "test:unit": "gulp mocha"
}
"nyc": {
    "check-coverage": true,
    "all": true,
    "extension": [
      ".js",
      ".jsx",
      ".ts",
      ".tsx"
    ],
    "include": [
      "src/**/!(*.test.*).[tj]s?(x)"
    ],
    "reporter": [
      "html",
      "lcov",
      "text",
      "text-summary"
    ],
    "report-dir": "docs/reports/coverage"
  }
```
`gulpfile.js` mocha relevant part
```javascript
const SRC_DIR = path.join(__dirname, 'src');
const SRC_FILES = path.join(SRC_DIR, '**', '*.[jt]s?(x)');
const TEST_FILES = path.join(SRC_DIR, '**', '*.test.[jt]s?(x)');
const MOCHA_CONFIG = {
    src: [
        TEST_FILES
    ],
    watchSrc: [
        SRC_FILES,
        TEST_FILES
    ],
    mocha: {
        // compilers: [
        //     'ts:ts-node/register',
        //     'tsx:ts-node/register'
        // ],
        require: [
            './tests/setup.js',
            'ignore-styles',
            'source-map-support/register'
        ]
    }
};
gulp.task('mocha', mocha(MOCHA_CONFIG));
```

`tsconfig.json`
```json
{
  "compilerOptions": {
    "baseUrl": "./",
    "rootDir": "./src",
    "outDir": "./build",
    "allowJs": true,
    "module": "commonjs",
    "target": "es5",
    "lib": ["es5", "es6", "dom"],
    "sourceMap": true,
    "inlineSourceMap": false,
    "inlineSources": false,
    "experimentalDecorators": true,
    "noUnusedParameters": true,
    "noUnusedLocals": true,
    "jsx": "react",
    "moduleResolution": "node"
  },
  "exclude": [
    "docs",
    "tests",
    "**/*.test.js",
    "**/*.test.jsx",
    "**/*.test.ts",
    "**/*.test.tsx",
    "tools",
    "gulpfile.js",
    "node_modules",
    "build",
    "typings/main",
    "typings/main.d.ts"
  ],
  "awesomeTypescriptLoaderOptions": {
    "useCache": true,
    "useBabel": true
  }
}
```

I've tried a number of approaches, some of them seem to be unable to use source maps and other fails due to `ts-node`/`tsc` errors.

With the above setup coverage produces results for all the files but they are incorrect for TS files most probably due to source maps not being used (i.e. report shows no coverage for lines which are comments and the numbers seems to be wrong as well).

With a number of variant approaches tried with no success one of the most commonly suggested is to add `"require": ["ts-node/register"]` to `nyc` configuration yet then I'm getting errors complaining about i.e. `gulpfile.js`, `docs/reports/coverage/lcov-report/prettify.js` and number of other JS files to be `not under 'rootDir'` which is correct yet it is not clear why `ts-node` tries to process all the files out of `src` even if they are excluded in `tsconfig.json` (still the configuration gets really complex).

## Forensic Information

**Operating System:** macOS 10.12.4
**Environment Information:** 
node v7.2.0
npm v3.10.9
nyc v11.0.1

**Comment by bradlord** [comment]
I'm also seeing this issue. I've created [a repository](https://github.com/bradlord/nyc-typescript-failing-example) to demonstrate the issue. In that repository, I created two files with identical contents. One has a .ts extension and the other has a .tsx extension. Interestingly, the coverage highlighting seems accurate in the .ts file, but not in the .tsx file:

![fooTs](https://user-images.githubusercontent.com/900379/28336164-eaf469ac-6bce-11e7-8c76-2a1a7da1d261.png)
![fooTsx](https://user-images.githubusercontent.com/900379/28336163-eaf02144-6bce-11e7-914e-e41483a34e53.png)

This suggests that there is some sort of issue with the source map, but I don't know enough to tell whether it's an issue with the generation of the source map or one of the libraries that are processing it.

I hope this helps.

**Comment by jvanoostveen** [comment]
I also see several of the second example coverage reports on .ts files as well. With files fully covered when using nyc 10.x.

**Comment by bcoe** [comment]
@bradlord running your example with `nyc@11.0.3` I seem to get identical output for bot the tsx and ts file.

I don't personally use TypeScript for any projects, so I'm not the best person in the world to debug issues; but perhaps we can loop in @JaKXz and @mohsen1 who have helped document the usage of TypeScript with nyc/istanbul.

I'll happily help land any updates to documentation we want to make, and any patches we come up with ; I'd also love to get regression tests in place for TypeScript if we can figure out a reasonable approach.

Any ways, as I'm not a TypeScript expert, any help improving support in nyc is greatly appreciated.

**Comment by bradlord** [comment]
I tracked down the difference between the TS and TSX coverage reporst to a [source map handling bug in ts-node.](https://github.com/TypeStrong/ts-node/issues/393) After upgrading to the latest version of ts-node, I'm at least seeing identical output for TS and TSX files.

**Comment by ciekawy** [comment]
upgrading to latest `nyc`, `ts-node` did not change anything. Yet just found that using `"target": "es5"` as `compilerOptions` in `tsconfig.conf` seems to solve the issue (need to confirm though if coverage report is fully correct).
Also if you need `es5` target in production there is still an option to use `TS_NODE_COMPILER_OPTIONS='{"target":"es6"}' nyc ...` i.e. in `package.json`'s `scripts` section

**Comment by bcoe** [comment]
@ciekawy interesting; nyc should understand ES6 syntax -- so not quite sure why compiling down to ES5 would make a difference.

@bradlord @jvanoostveen perhaps you can try the trick that @ciekawy came up with; and if that works we can document it in our README and typescript tutorial -- it seems like it would also be worthwhile to open up an issue on `ts-node`, tracked to this issue, I'd really love to get things behaving as expected, without much fiddling, between the two libraries.

**Comment by ciekawy** [comment]
I think it's otherwise as nyc/Istanbul seem to handle better ES6 than ES5. `"target": "es6"` which seems to work transpiles to ES6

**Comment by johan-gorter** [comment]
Hello, I am on the same team as @jvanoostveen. We already have `"target": "es5"`  in our `tsconfig.json`. 

The issue we are seeing with nyc 11.x seems to be a flaky one. I have 2 test-runs on identical source-files, one producing `Branches: 99.79% ( 8998/9017 )` and the other one producing `Branches: 100% ( 7537/7537 )`.

Hope this information is useful.

### Forensics
os: windows
nyc: 11.1.0
last known working version: 10.3.2
command: `cross-env TS_NODE_FAST=true nyc --check-coverage -- mocha --compilers ts:ts-node/register --reporter dot --recursive`

**Comment by bcoe** [comment]
@johan-gorter as indicated in the initial post, make sure that nyc has a `require` stanza that pulls in `"ts-node/register"` for you. This is required because the `--all` step functions by pre-instrumenting all of the files, and pulling out the coverage information.

If `nyc` itself is not the process performing the instrumentation, then there will be a discrepancy between the line count retrieved during the fake instrumentation applied by `--all` and the instrumentation applied when running tests that exercise the file.

tldr; if you're using `--all`, the `nyc` process needs to control the instrumentation of the files; feel free to share your `package.json` with me, and something might jump out.

**Comment by johan-gorter** [comment]
Thanks for pointing me in the right direction. I added the require statement to my package.json which contained the `all:true`.

We had problems with this require stanza before, producing vague error messages: `ts.Debug.assert(outputText !== undefined, "Output generation failed"`. I found out that this could be fixed this by adding `"**/*.d.ts"` to our excludes. (nyc seems to consider `file.d.ts` to have a `.ts` extension)

**Comment by johan-gorter** [comment]
The result now is that branches and functions are at 100% again. However, we now have 0.2% missing statement coverage and 0.21% missing line coverage. There seem to be at least two cases where we see missing coverage:

1) Import statements that import only interfaces from another file

```
    import { EditMode } from '../session';
```

2) Files that only contain an interface, like `point.ts` (Statements 0/1, Lines 0/1)

```
    export interface Point {
      x: number;
      y: number;
    }
```

When using `"all": false`, coverage is at 100%. Do you want me to file this a a new issue?
I did not observe any flaky behavior anymore by the way.

**Comment by bcoe** [comment]
@johan-gorter are you still seeing flaky behavior (coverage information that fluctuates)?

If this problem seems to have been corrected, I do recommend opening a new issue; the fact taht these files do not get exercised might be a different issue entirely; bonus points, perhaps we could figure out a minimum reproduction.

@ciekawy it sounds like you've got a working solution now, and we can probably close this ticket?

**Comment by johan-gorter** [comment]
I have not observed any flakyness anymore. I will try to find the time to create a minimum reproduction.

**Comment by bcoe** [comment]
@johan-gorter going to go ahead and close this, thanks for the reproduction.

**Comment by jbenoit2011** [comment]
By the way, do not know if this is related, but I noticed the `removeComments` option, in `tsconfig.json`, if enabled will remove Istanbul ignore tags and so actual coverage will be different from the expected one.

**Comment by bennyn** [comment]
There is a tutorial on how to use [Istanbul with TypeScript & mocha](https://istanbul.js.org/docs/tutorials/typescript/) but is there also a tutorial that explains how to use it with [Jasmine](https://jasmine.github.io)? 

I guess that it would be possible to instrument the compiled JavaScript code to get code coverage from TypeScript code?

**Comment by nevercast** [comment]
I found that using `"inlineSourceMap": true` in `tsconfig.json` and `inline-source-map` for `devtool` in my Webpack conf (Because I'm doing Typescript -> ES6 -> Babel -> ES5), seems to have fixed the alignment of my TS source in coverage. I am also requiring `source-map-support/register` with mocha's `-r` switch. 

Hopefully, this helps put anyone on the right track. I understand you may not be using the same setup as me (As I've Webpack in the middle) so our circumstance may be different.

**Comment by rinick** [comment]
I still have that issue on .ts code, I tried everything mentioned above, switching to es6 or es2017 using inlineSourceMap. none of them solved my issue.
when I run the test, 3 json files are generated in my .nyc_output. But their statementMap are different.  if nyc analyze one of the json first, the result is correct, but if it analyze the other file fist, and then try to merge the good one into it, the result would be broken.

**Comment by bcoe** [comment]
see: https://github.com/Microsoft/TypeScript/issues/24993 I would like for us to better document a happy path for TypeScript and nyc.

**Comment by JaKXz** [comment]
Since the S/O question has been answered, I'm going to close this issue for now. Thanks everyone for your help, and please feel free to add to the tutorial here: https://github.com/istanbuljs/istanbuljs.github.io

**Comment by ciekawy** [comment]
I am not tracking this heavily as this is not blocking me now. While the SO provides generic way to get the coverage working, there is another issue making the result non-deterministic - I figured out that this is due to the first file in the report is being randomly omitted which affects the overall coverage - I thought this is mentioned in another nyc ticket yet cannot find it now.

**Comment by bcoe** [comment]
@ciekawy in the past I've always tracked issues with non-determinism down to the merging of multiple coverage files. If you ever consistently bump into this issue, a repository with a minimal reproduction would be useful.
